logr.go

```diff
@@ -352,18 +359,33 @@ func (l Logger) Info(msg string, keysAndValues ...interface{}) {
 // triggered this log line, if present. The err parameter is optional
 // and nil may be passed instead of an error instance.
 func (l Logger) Error(err error, msg string, keysAndValues ...interface{}) {
-	if l.sink == nil {
+	if l.sink == nil || l.sink == discardLogSinkInstance {
 		return
 	}
 	if withHelper, ok := l.sink.(CallStackHelperLogSink); ok {
 		withHelper.GetCallStackHelper()()
 	}
+	l.sinkWithContextValues().Error(err, msg, keysAndValues...)
+}
+
+func (l Logger) sinkWithContextValues() LogSink {
+	if l.contextKV == nil {
+		return l.sink
+	}
+	if value := l.contextKV.sinkWithContextValues.Load(); value != nil {
+		// We have done the steps below already, no need to repeat
+		// them.
+		return value.(LogSink) // nolint: forcetypeassert
+	}
 	sink := l.sink
 	contextKeysAndValues := l.valuesFromContext()
 	if len(contextKeysAndValues) > 0 {
 		sink = sink.WithValues(contextKeysAndValues...)
```

Comment:
so this is assuming that the sink does the caching internally, right?  But that doesn't work, unless I am missing something, because the sink merges these with the regular WithValues().  As we clone sinks, the set of context-keys can change, and we don't know which ones to replace vs retain.

I think we need either:

a) Sink gets a new optional interface which is specifically "values from context" so any subsequent call replaces the set
or
b) cache values here in Logger and pass then through WithValues(), which will perform worse obviously.

Or am I misunderstanding?

Reply:
Let's see...

Without caching:
- logger := FromContext(): context and keys remembered in Logger, for example with key "foo" and value "bar" in context
- logger = logger.WithValues("foo", 1)
- logger.Info("hello"): "foo=bar" retrieved now, overwrites "foo=1"

With caching:
- actually the same

Let's look at an example where logging triggers the caching. Without caching:
- logger := FromContext(): context and keys remembered in Logger, for example with key "foo" and value "bar" in context
- logger.Info("hello"): logs "foo=bar"
- logger = logger.WithValues("foo", 1)
- logger.Info("hello"): "foo=bar" retrieved now, overwrites "foo=1"

With caching:
-  logger := FromContext(): context and keys remembered in Logger, for example with key "foo" and value "bar" in context
- logger.Info("hello"): logs "foo=bar" *and* stores that via WithValues
- logger = logger.WithValues("foo", 1): `setSink` creates a new `contextKV` without a sink in it
- logger.Info("hello"): re-applies the context key/value pairs and logs "foo=bar"


This needs unit tests, but I think the implementation already handles it so that context key/value pairs are always logged as if they were passed to `WithValues`, as documented.

Reply:
There are clearly several "interesting" interactions between modifying a logger through WithName and WithValues. In particular WithValues deserves more coverage, with scenarios for calling it before FromContext and after.

I've extended the benchmarks (and thus unit test) with that.

Reply:
There is one relevant difference: when using the new API to retrieve context values, that retrieval happens right before the log call, after any WithValues calls by the app.

```
logger := logger.WithContextHandler(... <returns X=something>... )
...
logger := logr.FromContext()
logger.WithValues("X", 1)
# implicit: logger.WithValues("X", something)
logger.Info("hello"
```

This means that context values always overwrite WithValues values.  It's not how the approach without this new API works, but I'm undecided whether that is enough reason to do something about this. Ultimately, key conflicts should better get avoided.

For now I just changed the definition of `WithContextHandler` to:
```
// WithContextHandlers extends the list of context handlers. When
// given a context through WithContext later, the Logger will invoke
// each handler and log the key/value pairs that they return as
// as if they had been passed to WithValues right before a log call.
```

You mentioned adding a LogSink API to store context key/value pairs. With such an API, the LogSink then could handle those pairs first (WithValue calls override them) or handle them last (WithValue calls get overridden, the current behavior).

Reply:
I checked out your PR to show you what I meant and I realized you've done something VERY clever.

You store the suplicate "derived" `sink`, but recreate it from the original sink - this is really subtle.  Additionally, the allocation of a new `contextKV` clearing the sink is also pretty hidden.  Both could use some comments :)

You raise 2 good questions, though.

1) Which is "more important" - key-vals from context or key-vals at the callsite?
2) When rendering, do "most important" values come first or last?  Go's JSON parsing says "last value wins": https://go.dev/play/p/ZE1xXilHHHg

Reply:
> Which is "more important" - key-vals from context or key-vals at the callsite?

I think we can define it however we like, because ultimately there shouldn't be any key conflicts, so it won't matter.

> When rendering, do "most important" values come first or last?

I would say last. Currently, the order is `WithValues` followed by parameters. I prefer to add the context values in the middle.

Reply:
Context values are added by the "admin" (main) and callsite values are added by the "user".  I sort of feel like context values coming last is correct (if a bit ugly).  I could be fine with either answer as long as it doesn't need EVEN MORE surface area to effect.

Reply:
I'm not sure about the "admin" vs. "user" distinction: the admin might end up configuring the logger to log certain context values (but even that isn't necessary anymore), but then the actual context values come from the code and thus the "user". It may be reasonable to allow the user some final control by letting the call site override the context.

But primarily this is driven by how this can be implemented: by injecting in the middle, we can do the caching via `WithValues`. That wouldn't work if we wanted the context to come first (before other `WithValues`) or last (after call site parameters).

Reply:
How do you want to proceed?  If you think it's important to let callsites override context, we need to have a different API than `sink.WithValues()`.  I think I am OK with it as-is because, as you said, there shouldn't be any key conflicts.  In fact, I wonder if it's worth the effort to do something like run a callback function if a duplicate key is detected or something.

Reply:
I prefer to keep it as-is, with no special support for detecting or reporting duplicate keys.